### PR TITLE
[Bugfix][Kernel] Promote BF16 causal-conv operands before accumulation

### LINKED ISSUE #52887: [Bug]: causal_conv1d Triton kernels round BF16 products before FP32 accumulation
## Environment

- vLLM current main: `5a4c8d99242e9e069b604d0e9b969e77f7dd501d`
- Reproduction base: `1a5b3bbe62e754990cc34a8e90799d34a7467ab9`
- GPU: NVIDIA H200, compute capability 9.0
- PyTorch: `2.11.0+cu130`
- CUDA: `13.0`
- Triton: `3.6.0`

## Bug

Both Triton causal-convolution kernels accumulate into FP32 but multiply the
input/history and weight while they are still BF16:

```python
acc += matrix_x * matrix_w
```

The affected sites on current main are in
`vllm/model_executor/layers/mamba/ops/causal_conv1d.py`, in
`_causal_conv1d_fwd_kernel` and `_causal_conv1d_update_kernel` (currently lines
455 and 1061). The BF16 product rounds before it reaches the FP32 accumulator.

The official `causal-conv1d==1.6.2.post1` CUDA implementation converts both
weights and inputs/history to `float` before the multiply-add. Its forward
kernel builds `float weight_vals[]` and `float x_vals[]`; its update kernel does
the same before `out_val += weight_vals[j] * x_vals[j]`.

The minimal correction is:

```diff
- acc += matrix_x * matrix_w
+ acc += matrix_x.to(tl.float32) * matrix_w.to(tl.float32)
```

at both sites. BF16 convolution history and output remain BF16. This does not
change recurrent-state dtype or routing semantics.

## Exact reproduction

A deterministic BF16 width-4 fixture compares the public APIs against a
sequential reference. The reference evaluates each BF16-operand multiply-add
in FP64 and rounds once to FP32 after every tap, which exactly models FP32 FMA:
BF16 operands have eight significant bits and the FP32 accumulator has 24, so
FP64 represents the intermediate exactly.

Coverage includes:

- packed prefill;
- ordinary one-token decode;
- packed varlen decode;
- speculative packed-varlen decode.

Before the two casts, all four cases fail the exact FP32-operand comparison
after their BF16-vs-FP32 sensitivity and dtype assertions pass. After the two
casts, all four are `torch.equal`. The complete causal-convolution and Qwen GDN
kernel regression is `260 passed`.

On a captured production Qwen GDN convolution over 131,072 values, current
vLLM is bit-for-bit equal to a BF16-product/FP32-accumulator reference. Its
relative L2 from official `causal-conv1d` is `0.193967%`. Promoting both operands
before multiplication reduces the official-kernel difference to `0.00009498%`
(one `0.000244140625` output-rounding difference).

## Performance consequence

The explicit promotion is not free. A paired H200 benchmark used identical
source except for the two casts, BF16 width-4 input/state/weights, SiLU, 50 ms
warmup, and 200 ms measurement per case. Values are microseconds, with
p20/median/p80 shown.

| Shape | Before | After | Median change |
|---|---:|---:|---:|
| D8192, T70 prefill | 146.938 / 157.088 / 170.035 | 175.750 / 181.408 / 197.882 | +15.48% |
| D8192, B1 decode | 6.816 / 6.944 / 7.104 | 7.072 / 7.232 / 7.456 | +4.15% |
| D1536, T70 prefill | 137.280 / 145.760 / 157.363 | 172.774 / 177.504 / 198.362 | +21.78% |
| D1536, B1 decode | 6.400 / 6.656 / 6.912 | 6.464 / 6.816 / 7.104 | +2.40% |

The prefill distributions do not overlap. The decode change is smaller and the
D1536 distributions overlap. Correctness should take precedence, but this cost
should be considered when optimizing the corrected kernel.

## Duplicate check

Searches for `causal_conv1d BF16 FP32 product rounding` and
`causal_conv1d FP32 operand` found no issue or open PR for this defect.

This is distinct from #51779, which concerns Q/K normalization and beta dtype
after convolution. It also does not duplicate #50226, which adds an SM103
long-prefill fast path. That PR's new generic path already casts its loaded
values and weights to FP32, but the existing fallback and update kernels remain
affected on current main.

AI assistance was used for the investigation, test construction, benchmark
automation, and drafting. The findings and proposed two-line change were
reviewed against the vLLM and official causal-conv1d sources.

### PR DESCRIPTION
**Pull request by maithilijoshi20**
## Purpose

Fixes #52887.

The BF16 causal-convolution Triton kernels accumulate in FP32, but their input/history values and weights were still multiplied as BF16 first. This changes those products to FP32 before accumulation, avoiding the extra BF16 rounding step.

State and output dtypes remain BF16.

## Testing

Added deterministic BF16 regression tests through the public APIs, covering regular decode, packed prefill, packed-varlen decode, and speculative decode.

```text
tests/kernels/mamba/test_causal_conv1d.py
169 passed

tests/models/kimi_k3/test_kda.py
50 passed, 6 skipped

### CODE REVIEW

**Comment by NolenLiang** [comment]
Data point in favor of this change, from the training/inference-consistency side. I compared vLLM's Triton causal conv against the causal-conv1d CUDA package (1.7.0 built from source, the kernel Megatron's memory-efficient Mamba path calls) on bf16 inputs with width 4 and 700 tokens, at Nemotron-H-8B (dim 10240) and Granite-4.0-H-350M (dim 1792) shapes. On main, 35% of the bf16 outputs differ (max abs 0.031); the CUDA output is within 0.001% of a correctly rounded fp64 reference, so the BF16 product rounding is the whole gap. With the two `.to(tl.float32)` promotions from this PR applied on top of my branch, the outputs are bitwise identical to the CUDA kernel for both shapes, and the choice of exp or division implementation makes no difference at bf16 output precision (checked 24 variants). The decode-equals-prefill tests I run for batch-invariant Mamba2 still pass with the change, since both kernels move together.

**Comment by Terrydaktal** [comment]
Additional R9700 validation of the same fix, from #57007. That PR duplicated the two operand promotions after the initial search missed this earlier PR; I am closing it in favor of this one and contributing the additional evidence here.

Environment: AMD Radeon AI PRO R9700 (`gfx1201`), ROCm 7.14, PyTorch `2.12.0+rocm7.14`, Triton `3.7.1+gitf0b55c07`.

| Check | Result |
| --- | --- |
| Original upstream kernel at `78eaaad2c7a7f8c7864d9c9c72772e3f9f2669d0`, with the added regression | 12 BF16/FP16 failures; all six FP32 controls passed |
| Same two operand promotions, complete convolution test file from #57007 | 182 passed; no failures, errors or skips |
| Applicable pre-commit checks | Passed |

The additional parameterized regression covers the prefill and update public APIs, BF16/FP16/FP32, widths 2/3/4, a masked channel tail and exact history updates, including the reserved null slot. It checks an exactly representable cancellation residual with zero tolerance: BF16 `1.0078125 * 1.0078125 - 1.015625` should retain `2**-14`, whereas the original kernel returns zero on this stack.

The [patch and regression are available in commit `edde11d`](https://github.com/Terrydaktal/vllm/commit/edde11df8b5d524db20f01b6d479219b42cdd934); please feel free to reuse the additional coverage. The regression is selected with `pytest tests/kernels/mamba/test_causal_conv1d.py -k preserves_product_precision -q`.

For this validation, an isolated runner imported that exact kernel source and test file into the pinned vLLM 0.28 environment, with `--noconftest` and pytest plugin autoload disabled. This validates the changed source on R9700; a complete current-main installation, CUDA/XPU execution and end-to-end throughput were not tested.
